### PR TITLE
Add mHC kernels for DeepSeek-V4

### PR DESCRIPTION
**Pull request by YangQun1**
This pull request adds support for new mHC (multi-Head Composition) XPU kernels and their benchmarking, along with improvements to the CMake build system to integrate these kernels and their dependencies. The most significant changes include the addition of new kernel source files, a benchmarking script for performance evaluation, and updates to build and include paths to support CUTLASS and related libraries.

**New mHC Kernel Support and Benchmarking:**

* Added new XPU kernel source files for mHC operations: `mhc_pre.cpp`, `mhc_post.cpp`, `hc_head_fused.cpp`, and `mhc_fused_post_pre.cpp`, enabling optimized implementations of DeepSeek V4 HC head fused and related kernels (`csrc/xpu/mhc/` and build integration). [[1]](diffhunk://#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR610-R613) [[2]](diffhunk://#diff-a0a834c32759e24e114ff64001b7592ad670d4cbbe13f0827a484a2692441eb0R1-R288)

* Introduced a new benchmarking script `benchmark/benchmark_mhc.py` to evaluate the performance of the new mHC kernels, including FLOPs, bandwidth, and arithmetic intensity calculations.

**Build System and Dependency Integration:**

* Updated `CMakeLists.txt` to append `VLLM_CUTLASS_FLAGS` to GPU compile flags and to include CUTLASS and related directories in the XPU include path if available, ensuring proper compilation and linkage of the new kernels. [[1]](diffhunk://#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR428) [[2]](diffhunk://#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR450-R458)# Essential Elements of an Effective PR Description Checklist

## Test Result

Correctness
- UT pass on B60

Performance:
- Measured performance on B60

## mhc_pre

| num_tokens | hidden_size | Latency (us) | TFLOPS | BW (GB/s) | Arith. Intensity |
|------------|-------------|--------------|--------|-----------|------------------|
|          1 |        4096 |       37.211 |  0.021 |    43.375 |            0.487 |
|         33 |        4096 |       54.247 |  0.478 |    53.962 |            8.866 |
|        128 |        4096 |       33.299 |  3.023 |   204.996 |           14.747 |
|        256 |        4096 |       57.962 |  3.473 |   208.398 |           16.667 |
|       1024 |        4096 |      224.060 |  3.594 |   194.581 |           18.471 |
|       2048 |        4096 |      471.068 |  3.419 |   181.763 |           18.811 |
|       4096 |        4096 |      902.018 |  3.571 |   188.104 |           18.985 |
|       8192 |        4096 |     1785.747 |  3.608 |   189.149 |           19.073 |
|      16384 |        4096 |     3575.851 |  3.603 |   188.479 |           19.118 |
|          1 |        7168 |       27.108 |  0.051 |   104.191 |            0.487 |
|         33 |        7168 |       45.382 |  1.001 |   112.834 |            8.869 |
|        128 |        7168 |       46.072 |  3.824 |   259.113 |           14.756 |
|        256 |        7168 |       87.719 |  4.016 |   240.806 |           16.679 |
|       1024 |        7168 |      398.348 |  3.538 |   191.378 |           18.486 |
|       2048 |        7168 |      812.634 |  3.468 |   184.237 |           18.826 |
|       4096 |        7168 |     1540.337 |  3.660 |   192.608 |           19.001 |
|       8192 |        7168 |     3017.861 |  3.736 |   195.705 |           19.089 |
|      16384 |        7168 |     5983.063 |  3.769 |   196.967 |           19.134 |

## mhc_post

| num_tokens | hidden_size | Latency (us) | TFLOPS | BW (GB/s) | Arith. Intensity |
|------------|-------------|--------------|--------|-----------|------------------|
|          1 |        4096 |        5.648 |  0.026 |    13.067 |            1.998 |
|         33 |        4096 |        6.254 |  0.778 |   389.487 |            1.998 |
|        128 |        4096 |        6.399 |  2.950 |  1476.403 |            1.998 |
|        256 |        4096 |       15.196 |  2.484 |  1243.384 |            1.998 |
|       1024 |        4096 |      192.965 |  0.782 |   391.674 |            1.998 |
|       2048 |        4096 |      384.014 |  0.786 |   393.628 |            1.998 |
|       4096 |        4096 |      766.032 |  0.788 |   394.654 |            1.998 |
|       8192 |        4096 |     1526.690 |  0.791 |   396.043 |            1.998 |
|      16384 |        4096 |     3058.126 |  0.790 |   395.429 |            1.998 |
|          1 |        7168 |        6.659 |  0.039 |    19.388 |            1.999 |
|         33 |        7168 |        5.375 |  1.584 |   792.705 |            1.999 |
|        128 |        7168 |        8.288 |  3.985 |  1993.775 |            1.999 |
|        256 |        7168 |       83.509 |  0.791 |   395.772 |            1.999 |
|       1024 |        7168 |      341.703 |  0.773 |   386.893 |            1.999 |
|       2048 |        7168 |      678.846 |  0.779 |   389.492 |            1.999 |
|       4096 |        7168 |     1351.444 |  0.782 |   391.293 |            1.999 |
|       8192 |        7168 |     2702.855 |  0.782 |   391.297 |            1.999 |
|      16384 |        7168 |     5398.523 |  0.783 |   391.818 |            1.999 |

## hc_head_fused

| num_tokens | hidden_size | Latency (us) | TFLOPS | BW (GB/s) | Arith. Intensity |
|------------|-------------|--------------|--------|-----------|------------------|
|          1 |        4096 |        5.718 |  0.023 |    53.010 |            0.432 |
|         33 |        4096 |        9.229 |  0.469 |   174.867 |            2.680 |
|        128 |        4096 |       11.392 |  1.473 |   483.224 |            3.048 |
|        256 |        4096 |       21.998 |  1.525 |   488.588 |            3.122 |
|       1024 |        4096 |      127.644 |  1.052 |   330.649 |            3.180 |
|       2048 |        4096 |      243.051 |  1.104 |   346.216 |            3.190 |
|       4096 |        4096 |      477.695 |  1.124 |   351.761 |            3.195 |
|       8192 |        4096 |      943.164 |  1.138 |   356.043 |            3.198 |
|      16384 |        4096 |     1862.385 |  1.153 |   360.479 |            3.199 |
|          1 |        7168 |        8.745 |  0.026 |    60.659 |            0.432 |
|         33 |        7168 |       15.109 |  0.501 |   186.917 |            2.680 |
|        128 |        7168 |       18.631 |  1.576 |   517.088 |            3.048 |
|        256 |        7168 |       37.453 |  1.568 |   502.201 |            3.122 |
|       1024 |        7168 |      224.313 |  1.047 |   329.268 |            3.180 |
|       2048 |        7168 |      424.920 |  1.106 |   346.558 |            3.190 |
|       4096 |        7168 |      841.572 |  1.116 |   349.417 |            3.195 |
|       8192 |        7168 |     1666.908 |  1.127 |   352.546 |            3.198 |
|      16384 |        7168 |     3324.926 |  1.130 |   353.350 |            3.199 |

## mhc_fused_post_pre

| num_tokens | hidden_size | Latency (us) | TFLOPS | BW (GB/s) | Arith. Intensity |
|------------|-------------|--------------|--------|-----------|------------------|
|          1 |        4096 |       20.528 |  0.045 |    80.625 |            0.564 |
|         33 |        4096 |       32.462 |  0.949 |   131.898 |            7.198 |
|        128 |        4096 |       38.739 |  3.086 |   311.810 |            9.896 |
|        256 |        4096 |       87.137 |  2.744 |   259.194 |           10.585 |
|       1024 |        4096 |      406.707 |  2.351 |   210.527 |           11.169 |
|       2048 |        4096 |      854.598 |  2.238 |   198.541 |           11.272 |
|       4096 |        4096 |     1676.238 |  2.282 |   201.506 |           11.325 |
|       8192 |        4096 |     3315.073 |  2.308 |   203.305 |           11.351 |
|      16384 |        4096 |     6632.132 |  2.307 |   203.007 |           11.364 |
|          1 |        7168 |       29.688 |  0.055 |    97.551 |            0.564 |
|         33 |        7168 |       48.647 |  1.109 |   153.942 |            7.202 |
|        128 |        7168 |       66.892 |  3.127 |   315.781 |            9.903 |
|        256 |        7168 |      176.717 |  2.368 |   223.485 |           10.594 |
|       1024 |        7168 |      736.174 |  2.273 |   203.372 |           11.178 |
|       2048 |        7168 |     1499.418 |  2.232 |   197.865 |           11.282 |
|       4096 |        7168 |     2897.152 |  2.311 |   203.859 |           11.334 |
|       8192 |        7168 |     5720.290 |  2.340 |   206.016 |           11.361 |
|      16384 |        7168 |    11392.115 |  2.350 |   206.650 |           11.374 |

## (Optional) Documentation Update

**BEFORE SUBMITTING, PLEASE READ <https://docs.vllm.ai/en/latest/contributing>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by wuxun-zhang** [comment]
@xinyu-intel @jikunshang please help review, thanks.

**Review by xinyu-intel — APPROVED** [review]

**Comment by YangQun1** [comment]
Hi @jikunshang, could you help to review

**Comment by YangQun1** [comment]
Hi @xiaolong-intel pls also take a look

**Review by jikunshang — COMMENTED** [review]
sorry for late review. mhc are gemm related kerenl, and we need leverage sycl-tla right? if so, I prefer not place these kernel into `csrc/xpu` folder, because:
1. it will pollute existing building flag option like [here](https://github.com/vllm-project/vllm-xpu-kernels/pull/425/changes#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR439)
2. xe2 and xe3p may have differenet impl.

**Comment by YangQun1** [comment]
> sorry for late review. mhc are gemm related kerenl, and we need leverage sycl-tla right? if so, I prefer not place these kernel into `csrc/xpu` folder, because:
> 
> 1. it will pollute existing building flag option like [here](https://github.com/vllm-project/vllm-xpu-kernels/pull/425/changes#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR439)
> 2. xe2 and xe3p may have differenet impl.

switched to follow other sycl-tla based kernels, to build a separated .so.

File: csrc/utils.h [hunk 1]
```diff
@@ -99,6 +99,20 @@ static inline bool force_xe_default_kernel() {
   return false;
 }
 
+// Control whether MHC kernels use the TF32 DPAS GEMM path (split-K).
+// Default: enabled (1). Set VLLM_MHC_USE_TF32=0 to force the vector path.
+static inline bool mhc_use_tf32() {
+  static const bool val = []() {
```
Comment (jikunshang) [inline csrc/utils.h]: do we need such a lambda here..

File: csrc/xpu/mhc/xe_2/hc_head_fused.cpp [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+  const int smem_size = static_cast<int>(
+      (HcHeadFusedOptKernel::NUM_RED * HcHeadFusedOptKernel::NUM_SG *
+           HcHeadFusedOptKernel::BLOCK_M +
+       HC * HcHeadFusedOptKernel::BLOCK_M) *
+      sizeof(float));
+
+  const auto sycl_block = compat::dim3(1, WG_SIZE, 1);
+  const auto sycl_grid = compat::dim3(static_cast<unsigned int>(num_wgs), 1, 1);
+
+#if !defined(SYCL_EXT_ONEAPI_WORK_GROUP_SCRATCH_MEMORY)
```
Comment (jikunshang) [inline csrc/xpu/mhc/xe_2/hc_head_fused.cpp]: where is this defined? looks like a oneapi compiler flag?

File: csrc/xpu/mhc/xe_2/hc_head_fused.cpp [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+      params);
+#else
+  compat::experimental::launch_properties launch_props{
+      sycl::ext::oneapi::experimental::work_group_scratch_size(smem_size),
+  };
+  compat::experimental::kernel_properties kernel_props{
+      sycl::ext::oneapi::experimental::sub_group_size<16>};
+  compat::experimental::launch_policy policy{
+      sycl_grid, sycl_block, launch_props, kernel_props};
+  auto event = compat::experimental::launch<
```
Comment (jikunshang) [inline csrc/xpu/mhc/xe_2/hc_head_fused.cpp]: please check whether this is safe as we have this issue before.
https://github.com/vllm-project/vllm-xpu-kernels/pull/298/changes
[internal jira](https://jira.devtools.intel.com/browse/CUTLASS9-606)

File: csrc/xpu/mhc/xe_2/mhc_pre.cpp [hunk 4]
```diff
@@ -0,0 +1,1082 @@
...
+
+  TORCH_CHECK(HC == 4, "mhc_pre: only hc_mult=4 is supported");
+  TORCH_CHECK(
+      fn_c.size(0) == HC3 && fn_c.size(1) == HC * H, "fn shape mismatch");
+  TORCH_CHECK(hc_scale.numel() == 3, "hc_scale must have 3 elements");
+  TORCH_CHECK(hc_base.numel() == HC3, "hc_base must have HC3 elements");
+
+  auto outer_shape = residual_c.sizes().slice(0, residual_c.dim() - 2).vec();
+  auto residual_flat = residual_c.view({-1, HC, H});
+  const int64_t N = residual_flat.size(0);
```
Comment (jikunshang) [inline csrc/xpu/mhc/xe_2/mhc_pre.cpp]: should we rename this? or use M_GEMM K_GEMM N_GEMM later.

File: csrc/utils.h (hunk 1 above)
Comment (YangQun1) [reply csrc/utils.h]: removed

File: csrc/xpu/mhc/xe_2/hc_head_fused.cpp (hunk 2 above)
Comment (YangQun1) [reply csrc/xpu/mhc/xe_2/hc_head_fused.cpp]: removed

File: csrc/xpu/mhc/xe_2/hc_head_fused.cpp (hunk 3 above)
Comment (YangQun1) [reply csrc/xpu/mhc/xe_2/hc_head_fused.cpp]: removed the event return

File: csrc/xpu/mhc/xe_2/mhc_pre.cpp (hunk 4 above)
Comment (YangQun1) [reply csrc/xpu/mhc/xe_2/mhc_pre.cpp]: renamed to num_tokens, and also change the MKN to have _GEMM suffix

File: csrc/xpu/mhc/xe_2/hc_head_fused.cpp (hunk 2 above)
Comment (jikunshang) [reply csrc/xpu/mhc/xe_2/hc_head_fused.cpp]: sorry, just curious. you can leave it if it's necessary.

Comment (YangQun1) [reply csrc/xpu/mhc/xe_2/hc_head_fused.cpp]: the definition should be inside compiler or sycl itself. before, it's to solve some compiler/sycl version mismatch issues (for example use the work_group_scratch_size api may have error in some older compiler). but in other kernels, I saw they only use the work_group_scratch_size for slm, so the macro is not necessary in vllm-xpu-kenrels latest env. so I removed it fully to align with other kernels. in mqa logits, we also have similar #if/else, we can remove them in later prs.

**Review by jikunshang — COMMENTED** [review]
@mayuyuace  please take another look.

**Review by mayuyuace — APPROVED** [review]

**Review by jikunshang — APPROVED** [review]
